packages/protocol/tipmanager/tipmanager.go

```diff
@@ -216,19 +174,171 @@ func (t *TipManager) selectTips(count int) (parents models.BlockIDs) {
-	return t.tips.Keys()
+func (t *TipManager) AllTips() (allTips []*scheduler.Block) {
+	allTips = make([]*scheduler.Block, 0, t.tips.Size())
+	t.tips.ForEach(func(_ models.BlockID, value *scheduler.Block) bool {
+		allTips = append(allTips, value)
+		return true
+	})
+
+	return
 }
 
 // TipCount the amount of tips.
 func (t *TipManager) TipCount() int {
 	return t.tips.Size()
 }
 
+// FutureTipCount returns the amount of future tips per epoch.
+func (t *TipManager) FutureTipCount() (futureTipsPerEpoch map[epoch.Index]int) {
+	futureTipsPerEpoch = make(map[epoch.Index]int)
+	t.futureTips.ForEach(func(index epoch.Index, commitmentStorage *memstorage.Storage[commitment.ID, *memstorage.Storage[models.BlockID, *scheduler.Block]]) {
+		commitmentStorage.ForEach(func(cm commitment.ID, tipStorage *memstorage.Storage[models.BlockID, *scheduler.Block]) bool {
+			futureTipsPerEpoch[index] += tipStorage.Size()
+			return true
+		})
+	})
+
+	return
+}
+
+// PromoteFutureTips promotes to the main tippool all future tips that belong to the given commitment.
+func (t *TipManager) PromoteFutureTips(cm *commitment.Commitment) {
+	t.evictionMutex.RLock()
+	defer func() {
+		t.evictionMutex.RUnlock()
+		t.Evict(cm.Index())
+	}()
+
+	if futureEpochTips := t.futureTips.Get(cm.Index()); futureEpochTips != nil {
+		if tipsForCommitment, exists := futureEpochTips.Get(cm.ID()); exists {
+
+			tipsToPromote := make(map[models.BlockID]*scheduler.Block)
+			tipsForCommitment.ForEach(func(blockID models.BlockID, tip *scheduler.Block) bool {
+				tipsToPromote[blockID] = tip
+				return true
+			})
+
+			tipsToNotPromote := set.NewAdvancedSet[models.BlockID]()
+			for _, tip := range tipsToPromote {
+				for _, parentID := range tip.Parents() {
+					if _, exists := tipsToPromote[parentID]; exists {
+						// Do not add this tips parent because this is not a tip anymore
+						tipsToNotPromote.Add(parentID)
+					}
+				}
+				// Remove all tips referenced by all future tips valid for this commitment
+				t.RemoveStrongParents(tip.ModelsBlock)
+			}
+
+			for tipID, tip := range tipsToPromote {
+				if !tipsToNotPromote.Has(tipID) {
+					t.addTip(tip)
+				}
+			}
+		}
+	}
+}
+
+// Evict removes all parked tips that belong to an evicted epoch.
+func (t *TipManager) Evict(index epoch.Index) {
+	t.evictionMutex.Lock()
+	defer t.evictionMutex.Unlock()
+
+	t.futureTips.Evict(index)
+}
+
 // endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 // region TSC to prevent lazy tips /////////////////////////////////////////////////////////////////////////////////////
 
+func (t *TipManager) addTip(block *scheduler.Block) (added bool) {
+	if !t.tips.Has(block.ID()) {
+		t.tips.Set(block.ID(), block)
+		// t.tipsConflictTracker.AddTip(block)
+		t.Events.TipAdded.Trigger(block)
+
+		// skip removing tips if a width is set -> allows to artificially create a wide Tangle.
+		if t.TipCount() <= t.optsWidth {
+			return
```

Comment:
```suggestion
			return true
```

Reply:
Tip was successfully added, but we just artificially don't remove parents to create wider tangle, so this should return true